include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+struct SAnisotropic
 {
-  // WARNING: Changed since GLSL, now arguments need to be normalized!
-  static Anisotropic<RayDirInfo> create(
-    const Isotropic<RayDirInfo> isotropic,
-    const float3 normalizedT,
-    const float normalizedB
-  )
-  {
-    Anisotropic<RayDirInfo> retval;
-    retval::Isotropic<RayDirInfo> = isotropic;
-    retval.T = normalizedT;
-    retval.B = normalizedB;
-    
-    const float3 V = retval.getDirection();
-    retval.TdotV = dot(V,retval.T);
-    retval.BdotV = dot(V,retval.B);
-
-    return retval;
-  }
-  static Anisotropic<RayDirInfo> create(const Isotropic<RayDirInfo> isotropic, const float3 normalizedT)
-  {
-    return create(isotropic,normalizedT,cross(isotropic.N,normalizedT));
-  }
-  static Anisotropic<RayDirInfo> create(const Isotropic<RayDirInfo> isotropic)
-  {
-    float2x3 TB = nbl::hlsl::frisvad(isotropic.N);
-    return create(isotropic,TB[0],TB[1]);
-  }
-
-  float3 getTangentSpaceV() {return float3(Tdot,BdotV,Isotropic<RayDirInfo>::NdotV);}
-  // WARNING: its the transpose of the old GLSL function return value!
-  float3x3 getTangentFrame() {return float3x3(T,B,Isotropic<RayDirInfo>::N);}
-
-  float3 T;
-  float3 B;
-  float3 TdotV;
-  float3 BdotV;
+    using ray_dir_info_type = RayDirInfo;
+    using scalar_type = typename RayDirInfo::scalar_type;
+    using vector3_type = typename RayDirInfo::vector3_type;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+    using isotropic_type = SIsotropic<RayDirInfo>;
+
+    // WARNING: Changed since GLSL, now arguments need to be normalized!
+    static SAnisotropic<RayDirInfo> create(
+        NBL_CONST_REF_ARG(isotropic_type) isotropic,
+        NBL_CONST_REF_ARG(vector3_type) normalizedT,
+        NBL_CONST_REF_ARG(vector3_type) normalizedB
+    )
+    {
+        SAnisotropic<RayDirInfo> retval;
+        retval.isotropic = isotropic;
+
+        retval.T = normalizedT;
+        retval.B = normalizedB;
+
+        retval.TdotV = nbl::hlsl::dot<vector3_type>(retval.isotropic.getV().getDirection(), retval.T);
+        retval.BdotV = nbl::hlsl::dot<vector3_type>(retval.isotropic.getV().getDirection(), retval.B);
+
+        return retval;
+    }
+    static SAnisotropic<RayDirInfo> create(NBL_CONST_REF_ARG(isotropic_type) isotropic, NBL_CONST_REF_ARG(vector3_type) normalizedT)
+    {
+        return create(isotropic, normalizedT, cross(isotropic.getN(), normalizedT));
+    }
+    static SAnisotropic<RayDirInfo> create(NBL_CONST_REF_ARG(isotropic_type) isotropic)
+    {
+        vector3_type T, B;
+        math::frisvad<vector3_type>(isotropic.getN(), T, B);
+        return create(isotropic, nbl::hlsl::normalize<vector3_type>(T), nbl::hlsl::normalize<vector3_type>(B));
+    }
+
+    vector3_type getT() NBL_CONST_MEMBER_FUNC { return T; }
+    vector3_type getB() NBL_CONST_MEMBER_FUNC { return B; }
+    scalar_type getTdotV() NBL_CONST_MEMBER_FUNC { return TdotV; }
+    scalar_type getBdotV() NBL_CONST_MEMBER_FUNC { return BdotV; }
```

Comment:
because you are using composition and not inheritance, in order to also make sure that `Isotropic<SAnisotropic<R> >==true` you need to provide the V and N getters, and a `create(V,N)` which hallucinates the rest of the tangent frame from frisvad